### PR TITLE
[prototype] Speed up `adjust_hue_image_tensor`

### PR DESCRIPTION
**Pull request by datumbox**
Related to #6818

Small performance improvement by making use of inplace ops where possible:
```
[-------------- adjust_hue_image_tensor cpu torch.float32 ---------------]
                         |  adjust_hue_image_tensor old  |     fn2 new    
1 threads: ---------------------------------------------------------------
      (16, 3, 400, 400)  |         467 (+- 13) ms        |  450 (+- 12) ms
      (3, 400, 400)      |          14 (+-  0) ms        |   14 (+-  0) ms
6 threads: ---------------------------------------------------------------
      (16, 3, 400, 400)  |         457 (+- 13) ms        |  451 (+-  9) ms
      (3, 400, 400)      |          18 (+-  0) ms        |   17 (+-  0) ms

Times are in milliseconds (ms).

[--------------- adjust_hue_image_tensor cuda torch.float32 --------------]
                         |  adjust_hue_image_tensor old  |      fn2 new    
1 threads: ----------------------------------------------------------------
      (16, 3, 400, 400)  |        2260 (+-  0) us        |  2200 (+-  1) us
      (3, 400, 400)      |         571 (+-  2) us        |   526 (+-  1) us
6 threads: ----------------------------------------------------------------
      (16, 3, 400, 400)  |        2261 (+- 10) us        |  2200 (+- 15) us
      (3, 400, 400)      |         570 (+- 19) us        |   525 (+-  4) us

Times are in microseconds (us).

[--------------- adjust_hue_image_tensor cpu torch.uint8 ----------------]
                         |  adjust_hue_image_tensor old  |     fn2 new    
1 threads: ---------------------------------------------------------------
      (16, 3, 400, 400)  |         483 (+- 20) ms        |  461 (+- 13) ms
      (3, 400, 400)      |          15 (+-  0) ms        |   15 (+-  0) ms
6 threads: ---------------------------------------------------------------
      (16, 3, 400, 400)  |         487 (+- 15) ms        |  479 (+- 17) ms
      (3, 400, 400)      |          19 (+-  0) ms        |   18 (+-  1) ms

Times are in milliseconds (ms).

[---------------- adjust_hue_image_tensor cuda torch.uint8 ---------------]
                         |  adjust_hue_image_tensor old  |      fn2 new    
1 threads: ----------------------------------------------------------------
      (16, 3, 400, 400)  |        2433 (+-  1) us        |  2365 (+-  0) us
      (3, 400, 400)      |         623 (+-  1) us        |   580 (+-  2) us
6 threads: ----------------------------------------------------------------
      (16, 3, 400, 400)  |        2436 (+-  6) us        |  2366 (+-  0) us
      (3, 400, 400)      |         622 (+- 21) us        |   581 (+-  2) us

Times are in microseconds (us).
```

cc @vfdev-5 @bjuncek @pmeier

### CODE REVIEW

File: torchvision/prototype/transforms/functional/_color.py [hunk 1]
```diff
@@ -204,11 +204,10 @@ def _rgb_to_hsv(image: torch.Tensor) -> torch.Tensor:
...
     mask_maxc_neq_r = maxc != r
     mask_maxc_eq_g = maxc == g
-    mask_maxc_neq_g = ~mask_maxc_eq_g
 
-    hr = (bc - gc).mul_(~mask_maxc_neq_r)
-    hg = (2.0 + rc).sub_(bc).mul_(mask_maxc_eq_g & mask_maxc_neq_r)
-    hb = (4.0 + gc).sub_(rc).mul_(mask_maxc_neq_g & mask_maxc_neq_r)
+    hg = rc.add(2.0).sub_(bc).mul_(mask_maxc_eq_g & mask_maxc_neq_r)
+    hr = bc.sub_(gc).mul_(~mask_maxc_neq_r)
+    hb = gc.add_(4.0).sub_(rc).mul_(mask_maxc_neq_r.logical_and_(mask_maxc_eq_g.logical_not_()))
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:210]: Changing the order of operations allows us to do more in-place ops.

In particular once `hg` is estimated, we can do an in-place on `bc` during the `hr` estimation. Then in the estimation of `hb` we can do inplace on `gc` and the logical masks.

File: torchvision/prototype/transforms/functional/_color.py [hunk 2]
```diff
@@ -234,7 +235,7 @@ def _hsv_to_rgb(img: torch.Tensor) -> torch.Tensor:
     a3 = torch.stack((p, p, t, v, v, q), dim=-3)
     a4 = torch.stack((a1, a2, a3), dim=-4)
 
-    return (a4.mul_(mask.to(dtype=img.dtype).unsqueeze(dim=-4))).sum(dim=-3)
+    return (a4.mul_(mask.unsqueeze(dim=-4))).sum(dim=-3)
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:238]: Unnecessary casting of mask. It's possible to do a multiplication with bools as we did previously for `mask_maxc_eq_g` etc.

File: torchvision/prototype/transforms/functional/_color.py [hunk 3]
```diff
@@ -217,14 +216,16 @@ def _rgb_to_hsv(image: torch.Tensor) -> torch.Tensor:
 
 def _hsv_to_rgb(img: torch.Tensor) -> torch.Tensor:
     h, s, v = img.unbind(dim=-3)
-    h6 = h * 6
+    h6 = h.mul(6)
     i = torch.floor(h6)
-    f = h6 - i
+    f = h6.sub_(i)
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:221]: Minor optimizations based on @pmeier's finding that mul is preferable to `*` for numbers. Also `h6` can be modified in-place as it's not reused.

**Review by datumbox — COMMENTED** [review]
Some clarification comments. Frankly I was expecting a more significant speed improvement given the amount of inplace ops that we could do here. The reason we don't see this is because `adjust_hue` is very computation heavy and one of our slowest transforms. These optimizations clip a few `ms` but those account for a small fraction of the total time.

**Review by pmeier — APPROVED** [review]
Thanks!
